### PR TITLE
Diff Encoding: compress even more by encoding diffs between frame token sequences

### PR DESCRIPTION
**Pull request by iejMac**
See experiments here - https://wandb.ai/iejmac/diff-encoding?workspace=user-iejmac

### CODE REVIEW

**Comment by iejMac** [comment]
Current status is I removed VQ altogether for now - encoding the diff on its own into some embedding should work before I try VQ. 

Main experiments on this branch are:
https://wandb.ai/iejmac/diff-encoding/runs/z2bqciix?workspace=user-iejmac
https://wandb.ai/iejmac/diff-encoding/runs/d6umrjas?workspace=user-iejmac

File: train/train.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        f_emb = enc(X)
+
+        # f, ppl, encodings = q(f_emb)
+        f = f_emb
+
+        logits = dec(e0, f)
+
+        # TODO: why does this matter???
+        true_logits = logits[:, :N_FRAME_TOKS]
+        # true_logits = logits[:, -N_FRAME_TOKS:]
```
Comment (iejMac) [inline train/train.py]: this makes training either completely not work or work and I currently have no clue why

**Comment by iejMac** [comment]
TODO:
- [x] multi-gpu training
- [x] scale normal run
- [x] better eval (on held out set, x1!=x0 accuracy)
- [x] figure out some of those weird things - why slicing a particular way matters etc.
- [x] instabilities (initialize transformers properly, make sure initial p dist is ~ uniform etc.)
- [x] weight tying
- [x] move on to VQ
- [x] add argparse and nicer launch scripts + model configs (improve interface)

**Comment by iejMac** [comment]
Instabilities TODO:
- [x] optimizations Hparams (lr, wd, eps)
- [x] init model params better

File: train/model.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+        t_embs = embs + p_embs
+
+        t_embs = t_embs.permute(1, 0, 2)  # NLD -> LND
+        c_embs = self.transformer(t_embs)
+        c_embs = c_embs.permute(1, 0, 2)  # LND -> NLD
+
+        # TODO: WHY DOES THIS MATTER
+        f = c_embs[:, :self.n_tokens]  # transformation is bottlenecked
+        # f = c_embs[:, -self.n_tokens:]  # transformation is bottlenecked
```
Comment (iejMac) [inline train/model.py]: need to understand this better, why does it matter

File: train/model.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        '''
+
+    def encode_diff(self, x):
+        return self.encoder(x)
+
+    def decode(self, x, f):
+        logits = self.decoder(x, f)
+        # TODO: why does this matter???
+        true_logits = logits[:, :N_FRAME_TOKS]  # for now only one frame
+        # true_logits = logits[:, -N_FRAME_TOKS:]
```
Comment (iejMac) [inline train/model.py]: same here, why does this matter at all

**Comment by iejMac** [comment]
TODO's:
- [x] Scale, larger models, more training etc. (optional, boring)
- [ ] Scale codebook
- [x] Weight tying
- [x] K-distance diffs (now we have model code correct so it might work)
- [ ] K-distance diff + AR (harder problem but stronger model)
- [x] Try AR posterior again (might be interesting for the latent loss component of this - latent code should also be predictable sequentially), try with both including x0, f in attn mask and also not
- [ ] Trajectories (might be interesting to try this with AR Posterior)
- [ ] Maybe try that distribution update idea again???
- [x] Might need to go MultiNode
- [x] Better checkpointing + Resuming failed runs


Notes:
- AR posterior is just too strong, model ignores everything. Maybe I can try AR but with a frame as the atomic element. Try this when investigating trajectories
- weight tying seems to slow down training...
- K-diff doesn't help, perplexity actually can't make progress...

**Comment by iejMac** [comment]
Gonna do a quick scaling plot. Notes:
- for all exp `n_heads = max(n_layers / 2, 8)`
- for all exp batch_size = 1024
- n_dynamics_tokens = 64 (=0.5 * n_frame_tokens)
- quantizer_embedding_width = 256

Scales:
- enc/dec 8L, 256 width = 13,669,888 p (short experiment to check if other changes had influence, will stop if not, compare with baseline)
- enc/dec 12L, 384 width = 44,031,232 p
- enc/dec 16L, 384 width = 58,226,944 p
- enc 24L/dec 16L, 384 width = 72,422,656 p
- enc 16L/dec 24L, 384 width = 72,422,656 p
- enc 24L/ dec 24L, 576 width = 193,018,240 p

Hypothesis: Models might be too deep, might need to go wider/use grad clipping

**Comment by iejMac** [comment]
Note: Adding MLP Projection layers around quantizer makes perplexity drop to 1.0 and latent_loss to go to 0 so codebook not used...
Linear works fine though. Find out why

File: train/model.py (hunk 3 above)
Comment (iejMac) [reply train/model.py]: doesn't matter as much now after fixes and my hypothesis (at least for decoder) is because the model is residual you want the contextualized embeddings of x0 to look like spatial embeddings (somewhat) so if you take the ones from the beginning you will get taht since the embeddings below are from the spatial embedding table and not your new (and constantly changing) diff codebook

**Comment by iejMac** [comment]
Another thought is that the way I'm currently doing it you could easily expand it to a rudimentary dynamic encoding method by just randomly choosing n_dynamics_tokens or choosing *all* n_dynamics_tokens from 1-N_FRAME_TOKENS or 1-N_FRAME_TOKENS/2 or something

**Comment by iejMac** [comment]
Next task - 43M (or like 20M) experiments until PPL of those is higher than our 13.6M models

**Comment by iejMac** [comment]
TODO:
- [x] Figure out better intialization method so avg_probs ~ uniform. Check how improves PPL and stuff
- [x] Add reinit mechanism

First point is not possible since that initial x1 = x0 local minimum forces codebook usage to suck

**Comment by iejMac** [comment]
TODO: Optimization params
- [ ] Impact of weight decay, try 1e-4 (maybe its diff here because you don't want to decay good codebook vectors)
- [ ] Impact of betas - try 0.9, 0.99
- [ ] LR schedule - warmup
- [ ] LR schedule - cosine decay
